presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java

```diff
@@ -359,4 +352,162 @@ public SetLookupSourceResult()
             return lookupSourceFutures;
         }
     }
+
+    private static class PartitionConsumption
+    {
+        private final int partitionNumber;
+        private final CompletableFuture<Void> requested;
+        private final CompletableFuture<Void> previousAdvanced;
+        private final CompletableFuture<LookupSourceFactory.LookupPartition> loaded;
+
+        public PartitionConsumption(
+                int partitionNumber,
+                CompletableFuture<Void> requested,
+                CompletableFuture<Void> previousAdvanced,
+                CompletableFuture<LookupSourceFactory.LookupPartition> loaded)
+        {
+            this.partitionNumber = partitionNumber;
+            this.requested = requested;
+            this.previousAdvanced = previousAdvanced;
+            this.loaded = loaded;
+        }
+    }
+
+    private class PartitionedConsumption
+    {
+        private final int consumers;
+        private final Session session;
+        private final PeekingIterator<PartitionConsumption> partitions = createPartitionConsumptions(spilledLookupSources.keySet());
+
+        private PartitionConsumption currentPartition;
+
+        PartitionedConsumption(int consumers, Session session)
+        {
+            this.consumers = consumers;
+            this.session = session;
+            advanceToNextPartition();
+        }
+
+        private PeekingIterator<PartitionConsumption> createPartitionConsumptions(Set<Integer> partitionNumbers)
+        {
+            List<PartitionConsumption> consumptions = partitionNumbers.stream()
+                    .map(PartitionedConsumption.this::createConsumption)
+                    .collect(Collectors.toList());
+            return peekingIterator(consumptions.iterator());
+        }
+
+        public int getConsumersCount()
+        {
+            return consumers;
+        }
+
+        private PartitionConsumption createConsumption(Integer partitionNumber)
+        {
+            CompletableFuture<Void> partitionRequested = new CompletableFuture<>();
```

Comment:
I would suggest cutting down on number of futures and dependencies between them, otherwise this code will be hard to understand and to follow (both when reading and while debugging).

For example , some (most?) of those "tasks" fired on future completions could be fired explicitly by caller methods. Like calling `release` could  could check whether that was the "last user of previous partition" and if so, call `LookupSourceFactory::releasePartition(...)` which would directly trigger unspilling next partition.